kernel/msg_q.c

```diff
@@ -239,14 +280,25 @@ int z_impl_k_msgq_get(struct k_msgq *msgq, void *data, k_timeout_t timeout)
 		if (pending_thread != NULL) {
 			SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_msgq, get, msgq, timeout);
 
-			/* add thread's message to queue */
-			__ASSERT_NO_MSG(msgq->write_ptr >= msgq->buffer_start &&
-					msgq->write_ptr < msgq->buffer_end);
-			(void)memcpy(msgq->write_ptr, pending_thread->base.swap_data,
+			if (pending_thread->base.pending_reason & _MSG_PREPEND) {
+				__ASSERT_NO_MSG(msgq->read_ptr >= msgq->buffer_start &&
+						msgq->read_ptr < msgq->buffer_end);
+				(void)memcpy(msgq->read_ptr, pending_thread->base.swap_data,
+			       msgq->msg_size);
+				msgq->read_ptr += msgq->msg_size;
+				if (msgq->read_ptr == msgq->buffer_end) {
+					msgq->read_ptr = msgq->buffer_start;
+				}
+			} else {
```

Comment:
As a slot in the message queue buffer has been made available, I think that we can simplify (and simultaneously correct) the prepend logic. Instead of copying to `read_ptr` (which if I am following things correctly overwrites a message), we can copy to `write_ptr` and then revert `read_ptr` to that. 

```
(void)memcpy(msgq->write_ptr, pending_thread->base.swap_data, msgq->msg_size);
msgq->read_ptr = msgq->write_ptr;
```

Reply:
Interesting idea, but if the msg queue is empty and `thread A` and `thread B` call `msgq_put` both, and then `thread C` call `msgq_get` we will miss a item.
Is it right?

Reply:
If the message queue is empty, then we are not in this if-block where (msgq->used_msgs > 0U). Instead, we would be farther down in the else clause.

Reply:
Sorry, I think I need to explain more detail.
if the msg queue is empty and `thread A` and `thread B` call `msgq_put`.
There are two items in msgq.
`Thread C` call `msgq_get` it will enter is block where (msgq->used_msgs > 0U).
In this condition, `msgq->read_ptr = msgq->write_ptr` will miss an item.